add troubleshooting-write-conflict (#3064)
pingcap/docs: troubleshoot-write-conflicts.md

---
title: TiDB Write Conflict In Optimistic Transaction Troubleshooting
summary: Learn about common errors and solutions of write conflict in optimistic transaction.
category: how-to
---

# TiDB Write Conflict In Optimistic Transaction Troubleshooting

This article is an introduction to the reason and solutions of write conflict in optimistic transaction.

> Before TiDB v3.0.8, optimistic transaction is the defalut transaction model in TiDB. The transaction doesn't check the confliction during the execution, it checks the confliction when committing the transaction. If the write conflict exists when committing the transaction, the transaction will be auto-retry in TiDB internal when TiDB has enabled the auto-retry and the retry count doesn't exceeded the limie. Then, TiDB will return the result of transaction execution to the client. Therefore, the write latency of TiDB cluster will be higher when there are a lot of write conflict exists.

Reviewer: limie -> limit

## The reason of write conflict

> TiDB use [Percolator](https://www.usenix.org/legacy/event/osdi10/tech/full_papers/Peng.pdf) transaction model as transaction implemention. And `percolator` is generally an implementation of 2PC. Please refer to [optimistic transaction document](/optimistic-transaction.md) for the specific 2PC process.

Reviewer 2: ```suggestion
TiDB implements its transactions by using the [Percolator](https://www.usenix.org/legacy/event/osdi10/tech/full_papers/Peng.pdf) transaction model. `percolator` is generally an implementation of 2PC. For the detailed 2PC process, see [TiDB Optimistic Transaction Model](/optimistic-transaction.md).
```

> After the client send a `COMMIT` request to TiDB, TiDB starts the 2PC process:

Reviewer 2: ```suggestion
After the client sends a `COMMIT` request to TiDB, TiDB starts the 2PC process:
```

1. TiDB chooses one key from all keys in the transaction as the primary key of the transaction.

> 2. TiDB sends `prewrite` requests to the TiKV regions which contain the all keys of the transaction.

Reviewer 2: ```suggestion
2. TiDB sends the `prewrite` request to all the TiKV Regions involved in this commit. TiKV judges whether all keys can preview successfully.
```

> 3. After all `prewrite` requests return successful result, go to the next step.

Reviewer 2: ```suggestion
3. TiDB receives the result that all `prewrite` requests are successful.
```

4. TiDB gets the `commit_ts` from PD.

> 5. TiDB sends `commit` request to the TiKV region which contain the primary key of the transaction. When TiKV executes `commit` request, TiKV checks the validity of the data and clear the lock that left in `prewrite` request.

Reviewer 2: ```suggestion
5. TiDB sends the `commit` request to the TiKV Region that contains the primary key of the transaction. After TiKV receives the `commit` request, it checks the validity of the data and clears the locks left in the `prewrite` stage.
```

> 6. After the `commit` request returns successfully, TiDB return success to the client.

Reviewer 2: ```suggestion
6. After the `commit` request returns successfully, TiDB returns success to the client.
```

> The write conflict occurs in the `prewrite` stage. When the transaction found a new version of the key (key.commit_ts > txn.start_ts), that's write conflict.

Reviewer: key.commit_ts -> data.commit_ts

> The write conflict occurs in the `prewrite` stage. When the transaction found a new version of the key (data.commit_ts > txn.start_ts), that's write conflict.

Reviewer 2: ```suggestion
The write conflict occurs in the `prewrite` stage. When the transaction finds that another transaction is writing the current key (`data.commit_ts` > `txn.start_ts`), a write conflict occurs.
```

> ## How to detect the write conflict exists in the cluster?

Reviewer 2: ```suggestion
## Detect write conflicts
```

You can check the following monitoring metrics under **KV Errors** in the **TiDB Grafana** panel:

* **KV Backoff OPS** indicates the count of error per second that return by TiKV.

    ![kv-backoff-ops](/media/troubleshooting-write-conflict-kv-backoff-ops.png)

    `txnlock` metric indicates the write-write conflict. `txnLockFast` metric indicates the read-write conflict.

* **Lock Resolve OPS** indicates the count of resolve locks per second which cause by transaction conflict:

    ![lock-resolve-ops](/media/troubleshooting-write-conflict-lock-resolve-ops.png)

    `not_expired` means the TTL of lock was not expired, Then the conflict transaction can't resolve locks until the TTL was expired.

>     `wait_expire` means the transaction needs to wait the lock to expired.

Reviewer: wait_expire -> wait_expired

    `expired` means the TTL of lock was expired, Then the conflict transaction can resolve this lock.

* **KV Retry Duration** indicates the duration of re-sends the KV request:

     ![kv-retry-duration](/media/troubleshooting-write-conflict-kv-retry-duration.png)

You can also use `[kv:9007]Write conflict` as the key word to search in TiDB log. The key word also indicates the write conflict exists in the cluster.

> ## How to resolve the write-conflict?

Reviewer 2: ```suggestion
## Resolve write conflicts
```
